### PR TITLE
Add single row subquery support

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -1,23 +1,22 @@
 /*
- * Licensed to Crate under one or more contributor license agreements.
- * See the NOTICE file distributed with this work for additional
- * information regarding copyright ownership.  Crate licenses this file
- * to you under the Apache License, Version 2.0 (the "License"); you may
- * not use this file except in compliance with the License.  You may
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
```
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: somehow the old header made it here?

File: blackbox/docs/sql/reference/value_expressions.txt [hunk 2]
```diff
@@ -185,3 +187,18 @@ optionally prefixed with ``ARRAY``::
...
+
+Scalar subquery
+===============
+
+A scalar subquery is a regular SELECT statement in parentheses that returns
+zero or one row with one column.
+
+If zero rows are returned it is treated as null value. In case more than one
+row is returned it is an error.
+
```
Comment (dobe) [inline blackbox/docs/sql/reference/value_expressions.txt]: should mention that it is not possible to refer to outside references

File: sql/src/main/java/io/crate/executor/transport/DelayedTask.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            @Override
+            public void onFailure(@Nonnull Throwable t) {
+                rowReceiver.fail(t);
+            }
+        });
+    }
+
+    @Override
+    public ListenableFuture<List<Long>> executeBulk() {
+        throw new UnsupportedOperationException("MultiPhaseTask doesn't support bulk operations");
```
Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/DelayedTask.java]: should mean DelayedTask

File: blackbox/docs/sql/reference/value_expressions.txt [hunk 4]
```diff
@@ -185,3 +187,18 @@ optionally prefixed with ``ARRAY``::
...
+.. note::
+
+    Scalar subqueries are restricted to SELECT statements and cannot be used in
+    other statements.
```
Comment (dobe) [inline blackbox/docs/sql/reference/value_expressions.txt]: pls also note that it only works in where and select list

File: sql/src/main/java/io/crate/executor/transport/SingleRowSingleValueRowReceiver.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import io.crate.core.collections.Row;
+import io.crate.operation.projectors.*;
+import io.crate.planner.Plan;
+
+import java.util.Set;
+
+/**
+ * RowReceiver expects to receive only one row and triggers a future with the value once completed
+ */
+class SingleRowSingleValueRowReceiver implements RowReceiver, CompletionListenable {
```
Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/SingleRowSingleValueRowReceiver.java]: other package to reuse later?

File: sql/src/main/java/io/crate/executor/transport/SubSelectSymbolReplacer.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import io.crate.planner.node.dql.join.NestedLoop;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Replaces all symbols in the querySpec that match the given selectSymbol once the futureCallback
+ * triggers with a Literal containing the value from the callback.
+ */
+class SubSelectSymbolReplacer implements FutureCallback<Object> {
```
Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/SubSelectSymbolReplacer.java]: other package to reuse later?

File: sql/src/main/java/io/crate/executor/transport/SingleRowSingleValueRowReceiver.java (hunk 5 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/SingleRowSingleValueRowReceiver.java]: probably, but I'd like to keep it package private until some other package needs it

**Pull request merged**